PhysicsTools/NanoAOD/plugins/PositionFromDetIDTableProducer.cc

```diff
@@ -0,0 +1,124 @@
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "DataFormats/NanoAOD/interface/FlatTable.h"
+#include "DataFormats/Common/interface/View.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHitContainer.h"
+#include "SimDataFormats/TrackingHit/interface/PSimHit.h"
+#include "CommonTools/Utils/interface/StringCutObjectSelector.h"
+
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCScintillatorDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCSiliconDetId.h"
+
+#include <DataFormats/GeometryVector/interface/GlobalPoint.h>
+#include <Geometry/CaloGeometry/interface/CaloGeometry.h>
+#include <Geometry/HGCalGeometry/interface/HGCalGeometry.h>
+#include <Geometry/Records/interface/CaloGeometryRecord.h>
+#include "Geometry/CommonDetUnit/interface/GlobalTrackingGeometry.h"
+#include "Geometry/Records/interface/GlobalTrackingGeometryRecord.h"
+#include "SimDataFormats/TrackingHit/interface/PSimHit.h"
+#include "DataFormats/CaloRecHit/interface/CaloRecHit.h"
+#include "DataFormats/HGCRecHit/interface/HGCRecHitCollections.h"
+
+#include <vector>
+#include <iostream>
+
+template <typename T>
+class PositionFromDetIDTableProducer : public edm::stream::EDProducer<> {
+public:
+  PositionFromDetIDTableProducer(edm::ParameterSet const& params)
+      : name_(params.getParameter<std::string>("name")),
+        doc_(params.getParameter<std::string>("doc")),
+        src_(consumes<T>(params.getParameter<edm::InputTag>("src"))),
+        cut_(params.getParameter<std::string>("cut"), true) {
+    produces<nanoaod::FlatTable>();
+  }
+
+  ~PositionFromDetIDTableProducer() override {}
+
+  void beginRun(const edm::Run&, const edm::EventSetup& iSetup) {
+      // TODO: check that the geometry exists
+      iSetup.get<CaloGeometryRecord>().get(caloGeom_);
+      iSetup.get<GlobalTrackingGeometryRecord>().get(trackGeom_);
+  }
+
+  GlobalPoint positionFromHit(const PCaloHit& hit) {
+    DetId id = hit.id();
+    return positionFromDetId(id);
+  }
+
+  GlobalPoint positionFromHit(const CaloRecHit& hit) {
+    return positionFromDetId(hit.detid());
+  }
+
+  GlobalPoint positionFromDetId(DetId id) {
+    DetId::Detector det = id.det();
+    int subid = (det == DetId::HGCalEE || det == DetId::HGCalHSi || det == DetId::HGCalHSc)
+                   ? ForwardSubdetector::ForwardEmpty
+                   : id.subdetId();
+    auto geom = caloGeom_->getSubdetectorGeometry(det, subid);
+
+    GlobalPoint position;
+    if (id.det() == DetId::Hcal) {
+      position = geom->getGeometry(id)->getPosition();
+    } else if (id.det() == DetId::HGCalEE || id.det() == DetId::HGCalHSi || id.det() == DetId::HGCalHSc) {
+      auto hg = static_cast<const HGCalGeometry*>(geom);
+      position = hg->getPosition(id);
+    }
+    else {
+        throw cms::Exception("PositionFromDetIDTableProducer") << "Unsupported DetId type";
+    }
+
+    return position;
+  }
```

Comment:
with #include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h

you can replace this with 
```
    hgcal::RecHitTools rhtools_;
    rhtools_.setGeometry(*caloGeom_);
    return rhtools_.getPosition(id);
```